### ISSUE TITLE
Unhandled CannotListenError in DNS query

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2581]
|<img alt="p1mrx's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| p1mrx reported|
|-|-|
|Trac ID|trac#2581|
|Type|defect|
|Created|2007-04-20 02:35:11Z|
|Branch|https://github.com/twisted/twisted/tree/names-listen-error-2581|

A user of my program sent this traceback to me.  It looks like it's possible, in rare cases, for a CannotListenError to occur deep inside the DNS query code.  I don't think it's correct for this error to be propagated to the caller of query().  The code works in 99.9% of cases, so I'm not sure what the source of the exception is in the first place.

Anyway, here's the traceback:

```
Traceback (most recent call last):
  File "twisted\python\context.pyo", line 37, in callWithContext 
  File "twisted\internet\selectreactor.pyo", line 139, in _doReadOrWrite
  File "twisted\internet\tcp.pyo", line 362, in doRead
  File "twisted\protocols\basic.pyo", line 149, in dataReceived
--- <exception caught here> ---
  File "dtella_dc.pyo", line 89, in lineReceived
  File "dtella_dc.pyo", line 390, in d_MyInfo
  File "dtella_dc.pyo", line 410, in removeLoginBlocker
  File "dtella_dc.pyo", line 456, in attachMeToDtella
  File "dtella.py", line 315, in addDCHandler
  File "dtella.py", line 194, in startConnecting
  File "dtella_dnslookup.pyo", line 122, in getConfigFromDNS
  File "twisted\names\common.pyo", line 24, in query
  File "twisted\names\common.pyo", line 131, in lookupText
  File "twisted\names\client.pyo", line 313, in _lookup
  File "twisted\names\client.pyo", line 229, in queryUDP
  File "twisted\names\dns.pyo", line 1158, in query
  File "twisted\names\dns.pyo", line 1104, in startListening
  File "twisted\internet\posixbase.pyo", line 388, in listenUDP
  File "twisted\internet\udp.pyo", line 84, in startListening
  File "twisted\internet\udp.pyo", line 92, in _bindSocket  
twisted.internet.error.CannotListenError: Couldn't listen on any:0: (10022, 'Invalid argument').
```

Attachments:

* [dns-listen.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fc8/fc839a0c77d2dd3fa4715619e786176d1f89e330/6046a18fec5b6ea91892c8f469645c153e0debcd.patch) (535 bytes) - added by p1mrx on 2007-09-09 23:25:27Z - Proposed patch

<details><summary>Searchable metadata</summary>

```
trac-id__2581 2581
type__defect defect
reporter__p1mrx p1mrx
priority__highest highest
milestone__ 
branch__branches_names_listen_error_2581 branches/names-listen-error-2581
branch_author__ 
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1177036511000000 1177036511000000
changetime__1199280529000000 1199280529000000
version__None None
owner__ 
cc__therve cc__exarkun cc__radix
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Can you offer some insight into how this error can occur?

It doesn't look possible to me. ;)

**Comment by twisted-trac** [comment]
|<img alt="p1mrx's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|p1mrx commented|
|-|-|

Replying to [exarkun](#note_1):
> Can you offer some insight into how this error can occur?
>
> It doesn't look possible to me. ;)

I don't know how to repeat it, but the important thing is that if you run the program on enough computers for a long enough time, eventually Windows will be Windows and spit out the mysterious error.  I just think that Twisted should catch this and turn it into a clean failure.

I encountered a somewhat related DNS problem (caused by Windows spewing random errors) In ticket [#2492](https://github.com/twisted/twisted/issues/2492).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

What kind of failure do you think makes sense here?

**Comment by twisted-trac** [comment]
|<img alt="p1mrx's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|p1mrx commented|
|-|-|

(5 months later, sorry...)

I think you should put a try/except around self.startListening() in dns.py, line 1158.  If any errors occur, have the query function return a defer.fail().

A function that returns a Deferred should *never* be raising exceptions, because that results in a breakdown of Twisted's asynchronous model.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @therve|
|-|-|

Let's wait for [#2492](https://github.com/twisted/twisted/issues/2492).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [21796]) Branching to 'names-listen-error-2581'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

Ready to review.

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_9"></a>|@Peaker commented|
|-|-|

There are a couple of extra space lines added before test_dns.py's first Test Class, but I am unaware of whether this fits convention.

Otherwise, it looks good to me.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to p1mrx|
|-|-|

I'd still like to know why this happens.

It's not practical to handle WSAEINVAL from every single Windows library call that gets made.  Until I know what's actually happening, I don't think I'll believe that handling it in random places doesn't improve the overall reliability of any Twisted-using application.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Ungg.

Here's something else, though.  listenUDP can reasonably be expected to raise CannotListenError sometimes.  If the branch catches _that_ instead of using a bare except, that's probably okay.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

(In [21879]) Do it.

Refs [#2581](https://github.com/twisted/twisted/issues/2581)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @therve|
|-|-|

Not sure how much I like ```test_listenOtherExceptions```.  Without actually knowing what exceptions can happen here, it's hard to decide if they should be synchronous or asynchronous.  I'd rather not pin us down to making them synchronous by encoding that requirement in a unit test.

How would you feel about just leaving that test out?  This issue is resolved just by handling the ```CannotListenError``` which is covered by the other test.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve commented|
|-|-|

(In [22160]) Remove suspicious test.

Refs [#2581](https://github.com/twisted/twisted/issues/2581)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve removed owner|
|-|-|

I thought it was the intended behavior, but sure, that's not necessary. I removed it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @therve|
|-|-|

Looks great, please merge

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve set status to `closed`|
|-|-|

(In [22248]) Merge names-listen-error-2581

Author: therve
Reviewer: exarkun
Fixes [#2581](https://github.com/twisted/twisted/issues/2581)

Handle CannotListenError in DNS client query code, and return a failure
instead.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
